plugin/plugins/wasmipld/wasmipld.go

```diff
@@ -0,0 +1,153 @@
+package wasmipld
+
+import (
+	"io/ioutil"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/multicodec"
+
+	mc "github.com/multiformats/go-multicodec"
+
+	"github.com/ipfs/go-ipfs/plugin"
+	"github.com/mitchellh/mapstructure"
+
+	wasmbind "github.com/aschmahmann/wasm-ipld/gobind"
+)
+
+// Plugins is exported list of plugins that will be loaded
+var Plugins = []plugin.Plugin{
+	&wasmipld{},
+}
+
+type wasmipld struct{}
+
+var _ plugin.PluginIPLD = (*wasmipld)(nil)
+var _ plugin.PluginIPLDADL = (*wasmipld)(nil)
+
+func (*wasmipld) Name() string {
+	return "ipld-wasmipld"
+}
+
+func (*wasmipld) Version() string {
+	return "0.0.1"
+}
+
+type WasmIPLDConfig struct {
+	Codecs []WasmIPLDCodecConfig
+	ADLs   []WasmIPLDADLConfig
+}
+
+type WasmIPLDCodecConfig struct {
+	Code     string
+	Encode   bool
+	Decode   bool
+	WasmPath string
+}
+
+type WasmIPLDADLConfig struct {
+	Name     string
+	WasmPath string
+}
+
+func (*wasmipld) Init(env *plugin.Environment) error {
+	config := env.Config
+	if config == nil {
+		return nil
+	}
+
+	var cfg WasmIPLDConfig
+	// Note: This dependency is just for convenience and is in our go.sum anyway
+	// It can go away if we upstream this into the main config instead of keeping it as a plugin
+	if err := mapstructure.Decode(config, &cfg); err != nil {
+		return err
+	}
+
+	registry = &wasmRegistry{}
+
+	for _, c := range cfg.Codecs {
+		wasm, err := ioutil.ReadFile(c.WasmPath)
```

Comment:
how much more annoying would it be to reference the wasm as a cid in the block store rather than an on-disk path?

Reply:
Wouldn't that lead to some kind of weird attacks, where you could create a format that have difference unrepeable sideeffects ?

(I guess we could configure the WASM interpreter to be deterministic and repeatable (no random, or io available, ...)).

Reply:
> how much more annoying would it be to reference the wasm as a cid in the block store rather than an on-disk path?

Not that much more, I'd like to be able to do that. Some things I think we'd want to do to enable this include:
1. A way to protect the WASM blocks from getting cleared by GC (should be pretty easy)
2. At least for now assert a 2MiB limit on WASM blobs make that they must be Raw blocks
3. Let the plugin get access to a blockservice